### ISSUE TITLE
Primitive projections, canonical structures, and SSReflect rewrite

### ISSUE THREAD

**Issue by vbgl** [coq/coq#10685]
The “rewrite” tactic from SSReflect (and probably other tactics that perform some matching) fails at inferring a canonical instance when the symbol that triggers the resolution is a primitive projection.

Here is a complete example. The issue arises when using SSReflect rewrite (standard rewrite is not affected), and when the instance (here boxed_bool, as opposed to boxed_nat) has primitive projections.

~~~
From Coq Require Import ssreflect.

Structure box (T: Type) := Box { sort :> Type ; projector : sort -> T }.

Lemma property T (B: box T) (x: B) :
  projector T B x = projector T B x.
Proof. exact eq_refl. Qed.

Record boxed_nat := BoxNat { unbox_nat : nat }.

Canonical box_nat_box : box nat :=
  {| projector := unbox_nat |}.

Check BoxNat 4 : sort _ (_ : box nat).

Goal forall n : boxed_nat, 0 <= unbox_nat n.
Proof.
  intros [n].
  rewrite property.
  apply le_0_n.
Qed.

Set Primitive Projections.

Record boxed_bool := BoxBool { unbox_bool : bool }.

Canonical box_bool_box : box bool :=
  {| projector := unbox_bool |}.

Check BoxBool true : sort _ (_ : box bool).

Goal forall b : boxed_bool, is_true (unbox_bool b || true).
Proof.
  intros [b].
  Fail rewrite property.
  change tt with tt.
  rewrite property.
  apply Bool.orb_true_r.
Qed.
~~~

Tested with versions 8.7.2, 8.8.2, 8.9.1, 8.10+β2 and master (6bb04f3): all tested versions exhibit the same behavior.

**Comment by gares** [comment]
The keyed match uses canonical values automatically, eg a projector (a constant recognized as such) matches any canonical values (such as unbox_bool). I guess there is a bug here https://github.com/coq/coq/blob/edd7519b6e1af6d62194f9f3dcc938534b86d036/plugins/ssrmatching/ssrmatching.ml#L505 it in a related piece of code (eg line 460 looks fishy).

**Comment by gares** [comment]
I'm on Vac until monday :-)

**Comment by vbgl** [comment]
Thanks for spotting a possible source for the bug. Regarding the code at line 460, I second Maxime’s comment (https://github.com/coq/coq/pull/391#discussion_r91284906).

Have nice holidays.

**Comment by gares** [comment]
> ```coq
> Canonical box_bool_box : box bool :=
>   {| projector := unbox_bool |}.
> ```

The problem comes from the discrepancy of the CS, as registered, and the canonical value, as present in the goal. Up there `unbox_bool` is a primitive projection, but is not applied. My guess is that the system picks the compatibility constant, and records in `Recordops.o_TCOMPS` that the canonical value has 0 arguments (so that if you have to solve `projector ?cs y = unbox_bool x` you solve by canonical structure `projector ?cs = unbox_bool` and by congruence `y = x`).

Here the goal features `unbox_bool x` in its "compact form `Proj(.., x)` and the number of arguments gets confusing for the code of ssr, since the CS as registered says 0, but the lemma wants one (the x of type B).

This also explain why `change tt with tt` makes it work (I've no clue why it expands... but it is good to know).

So I can probably fix this (the problem is that the various checks `n < na` and `n < np` are wrong in that scenario). But I'm not so sure this is the right way forward wrt the compat layer. IMO the term you declare  as a CS should fail to typecheck (is it a primproj? if so it must be applied). This I guess is also why I wrote `0` in the comment you point to (`Proj(a,b)` is not like `App(a,[|b|])` that counts one argumet, here `b` and `a` can't be ripped apart that is `a` is not a subterm of `Proj(a,b)`, while your example seem to suggest that some ports of the system think that `a'` (the compat projection) is a subterm...).

Not sure on how to proceed.

**Comment by gares** [comment]
> so that if you have to solve `projector ?cs y = unbox_bool x` you solve by canonical structure `projector ?cs = unbox_bool` and by congruence `y = x`

To clarify, this "unification step" when `unbox_bool x` is in compact form seems to suggest that `unbox_bool` is a subterm of `unbox_bool x` but it is not (unless you use `unbox_bool'`).

**Comment by vbgl** [comment]
> IMO the term you declare as a CS should fail to typecheck

Are you suggesting that there should not be any compatibility constant?

Moreover, if the projection is η-expanded, (as in “{| projector x := unbox_bool x |}”), then this declaration will be ignored by the canonical-instance resolution (no head constant). To overcome this issue, I must *name* that expanded projection and always use this wrapper instead of the primitive projection. This kind of defeats the point of primitive projections.

**Comment by gares** [comment]
No, I'm not suggesting anything. I believe your example is too contrived to draw conclusions, I can't tell what you want to do from that.

I'm questioning the fact that a primitive_projection can be mentioned unapplied. It seems needed for your example, but in general I can't tell. In particular you seem to want a class for projectors, for example to overload a generic concept such as "the first projection". But this is exactly what the `Proj` term constructor could do (pair an integer n and a term to mean the n-th field) and we could have syntax for it, eg `" _ .n"` (and ".n" wound simply not parse). In other words you seem to implement in userspace what primproj could (actually should) do in system space.

Can you be more explicit about the real use-case?

**Comment by vbgl** [comment]
The motivation behind this issue is to (always) enable primitive projections in math-comp. Many things break, e.g., some variations on “rewrite val_insubd” or “case: (val …)”. I could reproduce these failures in the simple example given above.

**Comment by ppedrot** [comment]
Let me join the chorus of complaints: relying on compatibility definitions for primitive projection is just asking for more trouble down the road. We have suffered enough already with this backwards-but-no-so-much compatible layer, and we should get rid of it at some point. It would be more profitable to design a first-class way to access primitive projections in userland.

**Comment by gares** [comment]
> is to (always) enable primitive projections in math-comp.

I understand that, but the canonical definition you declare *needs* the non-primitive projection.

Another way to put, the CS on `val` you minimized here should not even exist, IMO, if we go to primproj. I know it is then more work to port the code, so I understand the hope the compat layer manages it.

I'll try to fix the code, but to me it is not the way to go (eg, if we once an forall purge the compat layer in the future, we are going to need to port mathcomp again).

**Comment by gares** [comment]
FTR: `val` is implementing in userspace (with CS) what a generic `_ .1` projection would do, that it it works on any (registered) sigma-like type.

**Comment by vbgl** [comment]
It seems that a generic “first projection” is less general than “val”: one can declare a subType structure for data-types that are not records (for instance, if there is a data structure that nicely represents the quotient type).
